Flaky test: KafkaProtocolFilterReconcilerIT.shouldUpdateReferentAnnotationOnSecretModify fails due to timing issue

## Description

The `KafkaProtocolFilterReconcilerIT` integration test is flaky and intermittently fails in CI with an assertion error indicating a timing/race condition where an expected value change has not occurred.

## Failure Details

- **Test class**: `io.kroxylicious.kubernetes.operator.reconciler.kafkaprotocolfilter.KafkaProtocolFilterReconcilerIT`
- **Test method**: `shouldUpdateReferentAnnotationOnSecretModify`
- **Error**: `java.lang.AssertionError: Expecting actual: "AAAAAMgeUQQ" not to be equal to: "AAAAAMgeUQQ"`
- **Result**: `Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 54.96 s <<< FAILURE!`

## CI Run

Failed in workflow run: https://github.com/kroxylicious/kroxylicious/actions/runs/26142047804/job/76890204957 (PR #3984)

## Error Details

The test failed at line 191 with an assertion expecting a value to have changed, but it remained the same:

```
java.lang.AssertionError: 

Expecting actual:
  "AAAAAMgeUQQ"
not to be equal to:
  "AAAAAMgeUQQ"

	at io.kroxylicious.kubernetes.operator.reconciler.kafkaprotocolfilter.KafkaProtocolFilterReconcilerIT.shouldUpdateReferentAnnotationOnSecretModify(KafkaProtocolFilterReconcilerIT.java:191)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

## Analysis

This test expects that when a secret is modified, the referent annotation on the KafkaProtocolFilter should be updated to reflect the change. The assertion `assertThat(actual).isNotEqualTo(expected)` is checking that the annotation value has changed.

The failure indicates a race condition where:
1. The test modifies a secret
2. The test immediately checks if the referent annotation has been updated
3. The operator reconciliation has not yet completed by the time the assertion runs
4. The annotation still has the old value, causing the test to fail

This is a classic timing issue in Kubernetes operator tests where reconciliation is asynchronous and may take longer than expected in CI environments.

## Suggested Investigation

1. Review line 191 in KafkaProtocolFilterReconcilerIT.java to understand the timing/wait logic
2. Determine if there's an explicit wait for the reconciliation to complete
3. Consider using Awaitility or similar to poll for the expected annotation change with appropriate timeout
4. Check if the reconciler is slower in CI than local environments
5. Verify that the test waits for the secret modification to trigger and complete reconciliation before asserting

## Related Failure: shouldUpdateReferentAnnotationOnConfigMapModify

The same test class has another flaky test with identical symptoms:

- **Test method**: `shouldUpdateReferentAnnotationOnConfigMapModify`
- **Error**: `java.lang.AssertionError: Expecting actual: "AAAAADO/4Sk" not to be equal to: "AAAAADO/4Sk"`
- **Line**: 210 (vs line 191 for the Secret test)
- **CI Run**: https://github.com/kroxylicious/kroxylicious/actions/runs/26567207156/job/78264958456

This failure shows the exact same pattern - the test expects the referent annotation to update after a ConfigMap modification, but the reconciliation hasn't completed by the time the assertion runs.

Both tests (`shouldUpdateReferentAnnotationOnSecretModify` and `shouldUpdateReferentAnnotationOnConfigMapModify`) are testing the same feature with different resource types and both exhibit the same timing/race condition. This strongly suggests a common root cause in how the test waits for reconciliation to complete after modifying referenced resources.

--- a/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaprotocolfilter/KafkaProtocolFilterReconcilerIT.java
+++ b/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaprotocolfilter/KafkaProtocolFilterReconcilerIT.java
@@ -185,10 +185,22 @@ class KafkaProtocolFilterReconcilerIT {
                 .build());
 
         // then
-        assertAllConditionsTrue(filterOne);
-        String newChecksum = testActor.get(KafkaProtocolFilter.class, ResourcesUtil.name(filterOne)).getMetadata().getAnnotations()
-                .getOrDefault(Annotations.REFERENT_CHECKSUM_ANNOTATION_KEY, NO_CHECKSUM_SPECIFIED);
-        assertThat(newChecksum).isNotEqualTo(checksum);
+        // Editing the referent doesn't change the filter's metadata.generation, so an observedGeneration check
+        // alone is satisfied by the pre-edit state. Poll until the referent checksum changes (the reliable signal
+        // that the referent-triggered reconciliation completed) while confirming the conditions still hold (see #4018).
+        AWAIT.alias("referent checksum updated and conditions still true").untilAsserted(() -> {
+            var kpf = testActor.get(KafkaProtocolFilter.class, ResourcesUtil.name(filterOne));
+            assertThat(kpf.getStatus()).isNotNull();
+            KafkaProtocolFilterStatusAssert
+                    .assertThat(kpf.getStatus())
+                    .hasObservedGenerationInSyncWithMetadataOf(kpf)
+                    .conditionList()
+                    .singleElement()
+                    .isResolvedRefsTrue(kpf);
+            String newChecksum = kpf.getMetadata().getAnnotations()
+                    .getOrDefault(Annotations.REFERENT_CHECKSUM_ANNOTATION_KEY, NO_CHECKSUM_SPECIFIED);
+            assertThat(newChecksum).isNotEqualTo(checksum);
+        });
     }
 
     @Test
@@ -204,10 +216,22 @@ class KafkaProtocolFilterReconcilerIT {
                 .build());
 
         // then
-        assertAllConditionsTrue(filterOne);
-        String newChecksum = testActor.get(KafkaProtocolFilter.class, ResourcesUtil.name(filterOne)).getMetadata().getAnnotations()
-                .getOrDefault(Annotations.REFERENT_CHECKSUM_ANNOTATION_KEY, NO_CHECKSUM_SPECIFIED);
-        assertThat(newChecksum).isNotEqualTo(checksum);
+        // Editing the referent doesn't change the filter's metadata.generation, so an observedGeneration check
+        // alone is satisfied by the pre-edit state. Poll until the referent checksum changes (the reliable signal
+        // that the referent-triggered reconciliation completed) while confirming the conditions still hold (see #4018).
+        AWAIT.alias("referent checksum updated and conditions still true").untilAsserted(() -> {
+            var kpf = testActor.get(KafkaProtocolFilter.class, ResourcesUtil.name(filterOne));
+            assertThat(kpf.getStatus()).isNotNull();
+            KafkaProtocolFilterStatusAssert
+                    .assertThat(kpf.getStatus())
+                    .hasObservedGenerationInSyncWithMetadataOf(kpf)
+                    .conditionList()
+                    .singleElement()
+                    .isResolvedRefsTrue(kpf);
+            String newChecksum = kpf.getMetadata().getAnnotations()
+                    .getOrDefault(Annotations.REFERENT_CHECKSUM_ANNOTATION_KEY, NO_CHECKSUM_SPECIFIED);
+            assertThat(newChecksum).isNotEqualTo(checksum);
+        });
     }
 
     @Test

